ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,328 @@
+  -> TryAddTxs blk
+  -- ^ The value to interpret.
+  -> STM m (( [MempoolAddTxResult blk]
+            , [GenTx blk])
+           , [TraceEventMempool blk])
+runTryAddTxs stateVar = go [] []
+  where go processed evs (Done remaining) =
+          return ((reverse processed, remaining), reverse evs)
+        go processed evs (Trace tem tat') = do
+          go processed (tem:evs) tat'
+        go processed evs (WriteTx is tat') = do
+          writeTVar stateVar is
+          -- TODO: are intermediate states even needed? or could we stack them
+          -- until the end and only write once?
+          go processed evs tat'
+        go processed evs (Processed t tat') = do
+          go (t:processed) evs tat'
+
+-- | Craft a 'TryAddTxs' value representing the control flow for adding
+-- transactions to the mempool. See the documentation of 'runTryAddTxs' for some
+-- more context.
+pureTryAddTxs
+  :: ( LedgerSupportsMempool blk
+     , HasTxId (GenTx blk)
+     )
+  => LedgerConfig blk
+  -> (GenTx blk -> TxSizeInBytes)
+  -> [GenTx blk]
+  -> InternalState blk
+  -> TryAddTxs blk
+pureTryAddTxs cfg txSize = go
+  where
+    go [] _                      = Done []
+    go toAdd@(firstTx:toAdd') is
+      -- No space in the Mempool.
+      | let firstTxSize = txSize firstTx
+            curSize     = msNumBytes $ isMempoolSize is
+      , curSize + firstTxSize > getMempoolCapacityBytes (isCapacity is)
+      = Done toAdd
+
+      | otherwise
+      = maybeWriteState getValidationOutcome
+        where
+            (eVtx, vr)           = extendVRNew cfg firstTx txSize $ validationResultFromIS is
+            is'                  = internalStateFromVR vr
+            maybeWriteState      = if null (vrNewValid vr)
+                                   then id
+                                   else
+                                     -- Each time we have found a valid transaction,
+                                     -- we update the Mempool. This keeps our
+                                     -- STM transactions short, avoiding
+                                     -- repeated work.
+                                     --
+                                     -- Note that even if the transaction were invalid,
+                                     -- we could still write the state, because
+                                     -- in that case we would have that @is ==
+                                     -- is'@, but there's no reason to do that
+                                     -- additional write.
+                                     WriteTx is'
+
+            -- We only extended the ValidationResult with a single transaction
+            -- ('firstTx'). So if it's not in 'vrInvalid', it must be in
+            -- 'vrNewValid'.
+            getValidationOutcome = case eVtx of
+              Right vtx ->
+                assert (isJust (vrNewValid vr)) $
+                Trace (TraceMempoolAddedTx
```

Comment:
Style: there are some inconsistencies in how you indent the arguments on a multi-line application. EG here the arguments are indented by 1 space (or 2 from the opening paren). In the next `Trace`, they're indented by negative 1 space (or 0 from the opening paren). On line 206 below (`Just $\n...`) it's indented negative.

Generally, negative indent is unconventional and we shouldn't do that in this code base. The `StyleGuide.md` recommends 2 spaces (from the function name; regardless of the opening paren). The most important rule is to be consistent.